src/interop/js-interop/js-types.md

```diff
@@ -0,0 +1,212 @@
+---
+title: JS types
+description: Usage information about the core types in JS interop.
+---
+
+Dart values and JS values belong to separate language domains. When compiling to
+Wasm, they execute in separate *runtimes* as well. As such, you should treat JS
+values as foreign types. To provide Dart types for JS values,
+[`dart:js_interop`] exposes a set of [extension types] prefixed with `JS` called
+"JS types".
+
+Importantly, these types have compiler-specific [representation types]. This
+means that their runtime type will differ based on whether code is compiled to
+Wasm or JS. In order to interact with and examine these JS values, you should
+use [`external`] interop members.
+
+## Type hierarchy
+
+JS types form a natural type hierarchy:
+
+- Top type: `JSAny`, which is any non-nullish JS value
+  - Primitives: `JSNumber`, `JSBoolean`, `JSString`
+  - `JSSymbol`
+  - `JSBigInt`
+  - `JSObject`, which is any JS object
+    - `JSFunction`
+      - `JSExportedDartFunction`, which represents a Dart callback that was
+      converted to a JS function
+    - `JSArray`
+    - `JSPromise`
+    - `JSDataView`
+    - `JSTypedArray`
+      - JS typed arrays e.g. `JSUint8Array`
+    - `JSBoxedDartObject`, which allows users to box and pass Dart values
+      opaquely within the same Dart runtime
+
+You can find the definition of each type in the [`dart:js_interop` API docs].
+
+{% comment %}
+TODO (srujzs): Refer to this in the "interop type" section in syntax.
+TODO (srujzs): Should we add a tree diagram instead for JS types?
+{% endcomment %}
+
+## Conversions
+
+Since there are two separate domains, you will likely want to *convert* values
+between one domain to the other. For example, you may want to convert a Dart
+`String` into a JS string, which is represented by the JS type `JSString`. To do
+this, Dart supplies a number of `extension` members on various Dart types that
+will do the conversion for you, usually labeled as or prefixed with `toJS`.
+Similarly, you can convert JS values to Dart types. Dart supplies a number of
+members on various JS types to enable this conversion, usually labeled as or
+prefixed with `toDart`. They are often marked `external`, and their
+implementation is usually platform-dependent. Note that not all JS types have a
+conversion, and not all Dart types have a conversion.
```

Comment:
Thanks! My modifications after incorporating yours:

> to get the behavior and output you're expecting.

I think this part will likely confuse users more as it doesn't make clear what behavior and output they expect.

> This rewrite just adds a clause again just adding some context of why this is important.

I like the point of adding why users may want to do this, but ironically, `JSString` is a bad candidate since we allow you to write `String` on interop APIs. :) So, I opted to make this `List` <-> `JSArray` so that the motivation is more realistic.

> (or, to abstract the complexity of conversions for you)

While this is true for some conversions, users can't write the conversions themselves in many cases as they are implementation details that users don't have access to.

> If it's the latter, does the reader need to know about their implementation? If there's something the user needs to do when using these members that is dependent on whether the member's implementation marks it as external or not, can you explain why they need to know that here?

It's a good question. `external` is an overloaded keyword that serves many purposes, and I'm trying to reduce the confusion between the `external` members users can write and the ones that are written in the SDK. Some of the latter are *patched* members and not interop members i.e. they are implemented based on the compiler and don't have anything to do with JS interop. This may be a bit too much info for users so I shortened it to just mention that users shouldn't mistake this with interop members.